docs/examples/nmt/gnmt.md

```diff
@@ -0,0 +1,438 @@
+                                               src_max_len, tgt_max_len)
+        cache_dataset(data_train_processed, common_prefix + '_train')
+    data_val_processed = load_cached_dataset(common_prefix + '_val')
+    if not data_val_processed:
+        data_val_processed = process_dataset(data_val, src_vocab, tgt_vocab)
+        cache_dataset(data_val_processed, common_prefix + '_val')
+    data_test_processed = load_cached_dataset(common_prefix + '_test')
+    if not data_test_processed:
+        data_test_processed = process_dataset(data_test, src_vocab, tgt_vocab)
+        cache_dataset(data_test_processed, common_prefix + '_test')
+    fetch_tgt_sentence = lambda src, tgt: tgt.split()
+    val_tgt_sentences = list(data_val.transform(fetch_tgt_sentence))
+    test_tgt_sentences = list(data_test.transform(fetch_tgt_sentence))
+    return data_train_processed, data_val_processed, data_test_processed, \
+           val_tgt_sentences, test_tgt_sentences, src_vocab, tgt_vocab
+
+
+def get_data_lengths(dataset):
+    return list(dataset.transform(lambda srg, tgt: (len(srg), len(tgt))))
+
+
+data_train, data_val, data_test, val_tgt_sentences, test_tgt_sentences, src_vocab, tgt_vocab\
+    = load_translation_data(dataset=dataset, src_lang=src_lang, tgt_lang=tgt_lang)
+data_train_lengths = get_data_lengths(data_train)
+data_val_lengths = get_data_lengths(data_val)
+data_test_lengths = get_data_lengths(data_test)
+
+with io.open(os.path.join(save_dir, 'val_gt.txt'), 'w', encoding='utf-8') as of:
+    for ele in val_tgt_sentences:
+        of.write(' '.join(ele) + '\n')
+
+with io.open(os.path.join(save_dir, 'test_gt.txt'), 'w', encoding='utf-8') as of:
+    for ele in test_tgt_sentences:
+        of.write(' '.join(ele) + '\n')
+
+
+data_train = data_train.transform(lambda src, tgt: (src, tgt, len(src), len(tgt)), lazy=False)
+data_val = SimpleDataset([(ele[0], ele[1], len(ele[0]), len(ele[1]), i)
+                          for i, ele in enumerate(data_val)])
+data_test = SimpleDataset([(ele[0], ele[1], len(ele[0]), len(ele[1]), i)
+                           for i, ele in enumerate(data_test)])
+```
+
+## Create Sampler and DataLoader
+
+Now, we have obtained `data_train`, `data_val`, and `data_test`. The next step
+is to construct sampler and DataLoader. The first step is to construct batchify
+function, which pads and stacks sequences to form mini-batch.
+
+```python
+train_batchify_fn = btf.Tuple(btf.Pad(), btf.Pad(),
+                              btf.Stack(dtype='float32'), btf.Stack(dtype='float32'))
+test_batchify_fn = btf.Tuple(btf.Pad(), btf.Pad(),
+                             btf.Stack(dtype='float32'), btf.Stack(dtype='float32'),
+                             btf.Stack())
+```
+
+We can then construct bucketing samplers, which generate batches by grouping
+sequences with similar lengths.
+
+```python
+train_batch_sampler = FixedBucketSampler(lengths=data_train_lengths,
```

Comment:
Some explanation of where 1.2 comes from would be great

Reply:
this is empirically tuned.